Source/ASCollectionView.mm

```diff
@@ -2156,10 +2156,12 @@ - (void)layer:(CALayer *)layer didChangeBoundsWithOldValue:(CGRect)oldBounds new
   BOOL changedInNonScrollingDirection = (fixedHorizontally && newBounds.size.width != lastUsedSize.width) || (fixedVertically && newBounds.size.height != lastUsedSize.height);
 
   if (changedInNonScrollingDirection) {
-    [_dataController relayoutAllNodes];
-    [_dataController waitUntilAllUpdatesAreCommitted];
-    // We need to ensure the size requery is done before we update our layout.
+    // Because -invalidateLayout doesn't trigger any operations by itself, and we answer queries from UICollectionView using layoutThatFits:,
+    // we invalidate the layout before we have updated all of the cells. Any cells that the collection needs the size of immediately will get
+    // -layoutThatFits: with a new constraint, on the main thread, and synchronously calculate them. Meanwhile, relayoutAllNodes will update
+    // the layout of any remaining nodes on background threads (and fast-return for any nodes that the UICV got to first).
     [self.collectionViewLayout invalidateLayout];
```

Comment:
I think invalidating the collection layout should be done "in the future", inside `-[ASDataController _relayoutAllNodes]` [here](https://github.com/TextureGroup/Texture/blob/master/Source/Details/ASDataController.mm#L811). The reason is that there might be new layouts calculated before the size change, but not still stuck on the main serial queue and so not yet applied yet. Those layouts are invalid. But calling invalidating here is too soon and doesn't invalidate them.

So it looks like we need to change `-[ASDataController relayotuAllNodes]` to accept a block that get called on the main serial queue.

Reply:
@nguyenhuy Interesting; that sounds about right. This seems like a very important issue to fix. Do you have any ideas on how we would test it properly?

I'm a bit worried because I don't have time to dig into this right now. In fact, I haven't synced with upstream for almost 2 months, and I have to focus on some deadlines for Q4 and 2018 planning.

If you would be able to take over this diff, I'd certainly appreciate it. I bet this change would greatly help the Pinterest app, especially on iPad rotation.